frontend/src/queries/nodes/DataNode/dataNodeLogic.ts

```diff
@@ -0,0 +1,154 @@
+    }),
+    loaders(({ actions, values }) => ({
+        response: [
+            null as DataNode['response'] | null,
+            {
+                loadData: async () => {
+                    return (await query<DataNode>(values.query)) ?? null
+                },
+                loadNewData: async () => {
+                    if (!values.canLoadNewData || values.dataLoading) {
+                        return values.response
+                    }
+                    const diffQuery: EventsNode =
+                        values.response && values.response.results?.length > 0
+                            ? {
+                                  ...values.query,
+                                  after: values.response.results[0].timestamp,
+                              }
+                            : values.query
+                    const newResponse = (await query(diffQuery)) ?? null
+                    actions.highlightRows((newResponse?.results ?? []).map((r) => r.id))
+                    return {
+                        results: [...(newResponse?.results ?? []), ...(values.response?.results ?? [])],
+                        next: values.response?.next,
+                    }
+                },
+                loadNextData: async () => {
+                    if (!values.canLoadNextData || values.dataLoading) {
+                        return values.response
+                    }
+                    const diffQuery: EventsNode =
+                        values.response && values.response.results?.length > 0
+                            ? {
+                                  ...values.query,
+                                  before: values.response.results[values.response.results.length - 1].timestamp,
+                              }
+                            : values.query
+                    const newResponse = (await query(diffQuery)) ?? null
+                    return {
+                        results: [...(values.response?.results ?? []), ...(newResponse?.results ?? [])],
+                        next: values.response?.next,
+                    }
+                },
+            },
+        ],
+    })),
+    reducers(({ props }) => ({
+        dataLoading: [false, { loadData: () => true, loadDataSuccess: () => false, loadDataFailure: () => false }],
+        newDataLoading: [
+            false,
+            { loadNewData: () => true, loadNewDataSuccess: () => false, loadNewDataFailure: () => false },
+        ],
+        nextDataLoading: [
+            false,
+            { loadNextData: () => true, loadNextDataSuccess: () => false, loadNextDataFailure: () => false },
+        ],
+        autoLoadToggled: [
+            false,
+            // store the autoload toggle's state in localstorage, separately for each data node kind
+            { persist: true, storageKey: `queries.nodes.dataNodeLogic..autoLoadToggled.${props.query.kind}` },
```

Comment:
nit: Is the double dot here intended? I'm assuming you meant `queries.nodes.dataNodeLogic.autoLoadToggled.${props.query.kind}`.